src/connectors/refinement-list/connectRefinementList.js

```diff
@@ -155,137 +155,85 @@ export default function connectRefinementList(renderFn, unmountFn = noop) {
-            const facetValues = escapeFacetValues
-              ? escapeFacets(results.facetHits)
-              : results.facetHits;
-
-            const normalizedFacetValues = transformItems(
-              facetValues.map(({ value, ...item }) => ({
-                ...item,
-                value,
-                label: value,
-              }))
-            );
-
-            render({
-              items: normalizedFacetValues,
-              state,
-              createURL,
-              helperSpecializedSearchFacetValues,
-              refine: toggleRefinement,
-              isFromSearch: true,
-              isFirstSearch: false,
+    let toggleShowMore;
+
+    const createSearchForFacetValues = function(helper) {
+      return renderOptions => query => {
+        const { instantSearchInstance } = renderOptions;
+        if (query === '' && lastItemsFromMainSearch) {
+          // render with previous data from the helper.
+          renderFn(
+            {
+              ...this.getWidgetRenderState({
+                ...renderOptions,
+                results: lastResultsFromMainSearch,
+              }),
               instantSearchInstance,
-              isShowingMore,
+            },
+            false
+          );
+        } else {
+          const tags = {
+            highlightPreTag: escapeFacetValues
+              ? TAG_PLACEHOLDER.highlightPreTag
+              : TAG_REPLACEMENT.highlightPreTag,
+            highlightPostTag: escapeFacetValues
+              ? TAG_PLACEHOLDER.highlightPostTag
+              : TAG_REPLACEMENT.highlightPostTag,
+          };
+
+          helper
+            .searchForFacetValues(
+              attribute,
+              query,
+              // We cap the `maxFacetHits` value to 100 because the Algolia API
+              // doesn't support a greater number.
+              // See https://www.algolia.com/doc/api-reference/api-parameters/maxFacetHits/
+              Math.min(getLimit(this.isShowingMore), 100),
+              tags
+            )
+            .then(results => {
+              const facetValues = escapeFacetValues
+                ? escapeFacets(results.facetHits)
+                : results.facetHits;
+
+              const normalizedFacetValues = transformItems(
+                facetValues.map(({ value, ...item }) => ({
+                  ...item,
+                  value,
+                  label: value,
+                }))
+              );
+
+              const canToggleShowMore =
+                this.isShowingMore && lastItemsFromMainSearch.length > limit;
+
+              isFromSearch = true;
+
+              renderFn(
+                {
+                  ...this.getWidgetRenderState({
+                    ...renderOptions,
+                    results: lastResultsFromMainSearch,
+                  }),
+                  items: normalizedFacetValues,
+                  canToggleShowMore,
+                  canRefine: true,
+                  instantSearchInstance,
```

Comment:
this is an interesting pattern, but it makes sense, since getWidgetRenderState can't know about the state at this point. 

What if the widget renders again a second time when it's in search for facet values mode, won't it be overridden by the main items?

Reply:
I updated an existing test to check for the renderingOptions between renders https://github.com/algolia/instantsearch.js/pull/4549/commits/6ebeb4c76cd3c64d9a57460be0e50b53dab82fac

I ran the test on the previous implementation too, and got the same results (minus `toggleShowMore` that was `undefined` in `renderingOptions3` in the previous implementation, but it was a bug, now fixed).

Reply:
I don't think this tests an external render

Reply:
What do you mean by an external render ?

Reply:
a render that comes from another widget, for example a new search in search box triggering a render to refinementList. I want to make sure that the searched items still show, and it doesn't fall back to the items from hits

Reply:
I actually checked this now in the current version, and it behaves the same (query gets reset and no longer in search mode). It seems like that behaviour is on purpose, so this issue is to discuss separately!

Reply:
Ok, lets see that later on then.